src/ck-core/ck.C

```diff
@@ -548,13 +548,14 @@ void CkDeliverMessageFree(int epIdx,void *msg,void *obj)
 #if CMK_CHARMDEBUG
   CpdBeforeEp(epIdx, obj, msg);
 #endif    
+  const auto msgtype = UsrToEnv(msg)->getMsgtype();
```

Comment:
Why was this needed? The only reason I ask is because if it's needed due to AMPI modifying the message type, then wouldn't that cause a race?

Reply:
It's needed because the message can be freed in the call(). I'm not sure how, but it appears this is a common pattern when handling nokeep messages elsewhere in ck-core. AMPI doesn't modify charm message types at all.